hail/src/main/scala/is/hail/types/virtual/TStruct.scala

```diff
@@ -111,43 +109,72 @@ final case class TStruct(fields: IndexedSeq[Field]) extends TBaseStruct {
   }
 
-  override def insert(signature: Type, p: List[String]): (Type, Inserter) = {
-    if (p.isEmpty)
-      (signature, (a, toIns) => toIns)
-    else {
-      val key = p.head
-      val f = selfField(key)
-      val keyIndex = f.map(_.index)
-      val (newKeyType, keyF) = f
-        .map(_.typ)
-        .getOrElse(TStruct.empty)
-        .insert(signature, p.tail)
-
-      val newSignature = keyIndex match {
-        case Some(i) => updateKey(key, i, newKeyType)
-        case None => appendKey(key, newKeyType)
-      }
+  def insert(signature: Type, path: IndexedSeq[String]): (TStruct, Inserter) = {
 
-      val localSize = fields.size
+    val missing: Annotation =
+      null.asInstanceOf[Annotation]
 
-      val inserter: Inserter = (a, toIns) => {
-        val r = if (a == null || localSize == 0) // localsize == 0 catches cases where we overwrite a path
-          Row.fromSeq(Array.fill[Any](localSize)(null))
-        else
-          a.asInstanceOf[Row]
-        keyIndex match {
-          case Some(i) => r.update(i, keyF(r.get(i), toIns))
-          case None => r.append(keyF(null, toIns))
-        }
+    def updateRow(typ: TStruct, idx: Int)(f: Inserter)(a: Annotation, v: Any): Annotation =
+      a match {
+        case r: Row =>
+          r.update(idx, f(r.get(idx), v))
+        case _ =>
+          val arr = new Array[Any](typ.size)
+          arr.update(idx, f(missing, v))
+          Row.fromSeq(arr)
       }
-      (newSignature, inserter)
+
+    def insertField(typ: TStruct)(f: Inserter)(a: Annotation, v: Any): Annotation = {
+      val arr = new Array[Any](typ.size + 1)
+      a match {
+        case r: Row =>
+          for (i <- 0 until typ.size) {
+            arr.update(i, r.get(i))
+          }
+        case _ =>
+      }
+      arr(typ.size) = f(missing, v)
+      Row.fromSeq(arr)
     }
+
+    val (newType, inserter) =
+      path
+        .view
+        .scanLeft((this, identity[Type] _, identity[Inserter] _)) {
+          case ((parent, _, _), name) =>
+            parent.selfField(name) match {
+              case Some(Field(name, struct: TStruct, idx)) =>
+                (
+                  struct,
+                  typ => parent.updateKey(name, idx, typ),
+                  updateRow(parent, idx)
+                )
+              case Some(Field(name, _, idx)) =>
+                (
+                  TStruct.empty,
+                  typ => parent.updateKey(name, idx, typ),
+                  updateRow(parent, idx)
+                )
+              case None =>
+                (
+                  TStruct.empty,
+                  typ => parent.appendKey(name, typ),
+                  insertField(parent)
+                )
+            }
+        }
+        .foldRight((signature, ((_, toIns) => toIns): Inserter)) {
+          case ((_, insertField, transform), (newType, inserter)) =>
+            (insertField(newType), transform(inserter))
+        }
+
+    (newType.asInstanceOf[TStruct], inserter)
   }
```

Comment:
After merging Patrick's changes to the parser, I had a merge conflict in `MatrixExplodeRows` as that node used `Let` to compute the type. I decided to simplify how that node computed its type. That sent me down this rabbit hole.

The main change is to remove `fieldOption` and `insert` from `IR` - these operations only make sense on `TStruct`. I then re-wrote how the new-type and inserter was generated to be iterative rather than recursive.